### ISSUE TITLE
boehm_gc: test failure

### ISSUE THREAD

**Issue by alaviss** [haikuports/haikuports#3851]
Haiku x86_64 hrev53112

```rst
================================
   gc 8.0.4: ./test-suite.log
================================

# TOTAL: 16
# PASS:  14
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: gctest
============

pthread_mutex_unlock failed
Abort
mmap(PROT_NONE) failed
Abort
FAIL gctest (exit status: 149)

FAIL: disclaim_weakmap_test
===========================

GC Warning: Memory unmapping is disabled as incompatible with MPROTECT_VDB
Assertion failure, line 132: 0 == err
Assertion failure, line 132: 0 == err
Assertion failure, line 132: 0 == err
Assertion failure, line 132: 0 == err
Assertion failure, line 132: 0 == err
Assertion failure, line 132: 0 == err
FAIL disclaim_weakmap_test (exit status: 70)
```

**Comment by ivmai** [comment]
\+ @paurkedal (the author of disclaim_weakmap_test)

**Comment by paurkedal** [comment]
Is the failure consistently reproducible, and is it immediate?  What are the values of `err` in terms of Haiku `errno.h` symbols?

**Comment by ivmai** [comment]
The bug is sometimes reproducible This seems to be the 1st time after the upgrade to gc-8.0.4 on Apr 12 (this release added the test).

**Comment by ivmai** [comment]
This crash looks like to be similar to other crashes on Haiku which were workarounded:
* https://github.com/ivmai/bdwgc/commit/42ff92661
* https://github.com/ivmai/bdwgc/commit/8a455421b

**Comment by ivmai** [comment]
> What are the values of err in terms of Haiku errno.h symbols?

"0 == err" just mean pthread_mutex_lock() failed. But the errno code is not printed. (If pthread_mutex_lock is based on sem_wait then it might probably be of EACCES value.)

**Comment by waddlesplash** [comment]
Uh, if there's a bug in Haiku, was a ticket ever filed for it?

**Comment by ivmai** [comment]
> Uh, if there's a bug in Haiku, was a ticket ever filed for it?

I guess no, it is a matter of making minimal test case not involving the boehm_gc. See https://github.com/haikuports/haikuports/pull/2215#issuecomment-367500693

**Comment by ivmai** [comment]
> FAIL: gctest
> ============
>
> pthread_mutex_unlock failed
> Abort
> mmap(PROT_NONE) failed
> Abort

This fails are unclear to me. errno values are needed.

**Comment by paurkedal** [comment]
Note that functions from the pthread library return the error number, rather than setting `errno`.

**Comment by alaviss** [comment]
Can we revert to old boehm? This newer version breaks every users of the gc.

**Comment by ivmai** [comment]
As a workaround, we could pass --disable-munmap to configure

**Comment by Begasus** [comment]
Still failure/crash on the test (even with workaround), adding debug logs from todays testrun (without the workaround).
[lt-gctest-106331-debug-28-07-2024-14-38-26.report.txt](https://github.com/user-attachments/files/16404093/lt-gctest-106331-debug-28-07-2024-14-38-26.report.txt)
[lt-gctest-106369-debug-28-07-2024-14-38-14.report.txt](https://github.com/user-attachments/files/16404094/lt-gctest-106369-debug-28-07-2024-14-38-14.report.txt)
[lt-gctest-106374-debug-28-07-2024-14-38-11.report.txt](https://github.com/user-attachments/files/16404095/lt-gctest-106374-debug-28-07-2024-14-38-11.report.txt)

**Comment by waddlesplash** [comment]
I've discovered the problem: it's due to this chunk of code: https://github.com/ivmai/bdwgc/blob/6e5d2222e26356f3d2bdc05014b8019df082df6d/os_dep.c#L2297

posix_memalign just requests memory from the regular malloc() heap. When the GC unmaps this, the malloc() implementation doesn't expect that, and we get crashes.

Probably the GC should just use regular mmap. Why was this alternative method written in the first place? It looks like it was checked in as part of a Mono implementation. I tried ripping it out, but interestingly the resulting `gctest` trips some asserts in Haiku's kernel, which I'm working on solving (actually it looks like a pretty good stress-test for the page-protections system.)

However, it's also very slow for reasons I don't yet understand. CPU usage overall is very low; I got a brief report from the scheduling analyzer, and while pthread lock waits look rather high, it doesn't seem high enough to explain the slowness. Signal delivery problems, perhaps? I think I saw at least one abort() with that as the cause.

**Comment by ivmai** [comment]
> Probably the GC should just use regular mmap.

Right, good catch! If memory could be obtained from OS using mmap(), then posix_memalign() should not be used. I suppose the logic could be like: if USE_MUNMAP is defined then use mmap() else use posix_memalign(). I'm ready to process the PR;)

> Why was this alternative method written in the first place? It looks like it was checked in as part of a Mono implementation.

Probably. Before boehmgc-8.0.0 the memory unmapping feature was off by default.

**Comment by waddlesplash** [comment]
Seems sometimes it also fails with:
```
List reversal produced incorrect list - collector is broken
```

**Comment by waddlesplash** [comment]
Another different output:
```
Running on HAIKU/X86_64 target
Supported VDBs: manual mprotect
Switched to incremental mode
Emulating dirty bits with mprotect/signals
Wrong finalization data - collector is broken
Test failed
Wrong finalization data - collector is broken
Test failed
Wrong finalization data - collector is broken
Test failed
Wrong finalization data - collector is broken
Test failed
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
GC Warning: Finalization cycle involving 0x2081c0
Lost a node at level 16 - collector is broken
Test failed
Child process failed, pid= 20057, status= 0x1500
Test failed
Kill Thread
```

**Comment by waddlesplash** [comment]
I've pushed my WIP changes to https://github.com/waddlesplash/bdwgc/commit/aed356cb0676ce1f6d4e08735afe4b222c7b19ba, so others can test them and see what happens at least. All tests pass except `gctest`, which randomly fails with different outputs as above. I tried changing more configuration options, using `-O0`, etc. and nothing worked.

It's possible I got something wrong in my changes, of course, but I don't know what.

**Comment by waddlesplash** [comment]
On some failures I get apparently-bogus (nearly-NULL) pointers. Missing locking somehow?
```
Running on HAIKU/X86_64 target
Supported VDBs: manual mprotect
Switched to incremental mode
Emulating dirty bits with mprotect/signals
Lost a node at level 16 - collector is broken
Test failed
Unexpected segmentation fault outside heap at 0xd1
Unexpected segmentation fault outside heap
Lost a node at level 16 - collector is broken
Test failed
Unexpected segmentation fault outside heap at 0x7f5e441e5000
Unexpected segmentation fault outside heap
Unexpected segmentation fault outside heap at 0xd1
Unexpected segmentation fault outside heap
Wait for child process failed
Test failed
Child process failed, pid= 649, status= 0x1500
Test failed
Kill Thread
```

**Comment by ivmai** [comment]
> Another different output

Output does not matter actually, each message contains "the collector is broken".

**Comment by ivmai** [comment]
> I've pushed my WIP changes to https://github.com/waddlesplash/bdwgc/commit/aed356cb0676ce1f6d4e08735afe4b222c7b19ba

The change affects many aspects at once and the result is "broken collector", I suggest to improve the collector for Haiku one by one (e.g. start with removing GC_haiku_get_mem in favor of mmap-based memory allocation, then try enable munmap).

**Comment by ivmai** [comment]
> In incremental mode, this happens

Could you please clarify if gctest fails in incremental mode in boehm_ gc-8.2.x (with `--disable-munmap`)?

**Comment by ivmai** [comment]
> Another different output

To simplify debugging, I think you could first compile and run tests in single-threaded mode (by passing `--disable-threads`).

**Comment by waddlesplash** [comment]
> The change affects many aspects at once and the result is "broken collector", 

I think it was broken even before I made any changes (even without memory unmapping), but yes, I can try with just the original boehm-gc source without modifications and the `--disable-` flags you suggest and see what happens.

**Comment by waddlesplash** [comment]
Using an unmodified `gc-8.2.8`. With only `--disable-munmap` (threads still enabled), it just crashes at once with "pthread_mutex_unlock failed", so I applied just the change to use `pthread_mutex_init` in the child from my patchset. It then failed with:
```
Switched to incremental mode
Emulating dirty bits with mprotect/signals
Unexpected bus error or segmentation fault at 0x117feda72a20
Unexpected bus error or segmentation fault
Abort
Kill Thread
```

So I adjusted `gcconfig.h` to disable the `mprotect` VDB. This failed even earlier:
```
Child process failed, pid= 6252, status= 0x1500
Test failed
Abort
Kill Thread
```

I then reconfigured with `--disable-munmap --disable-threads`. This failed again very early on:
```
GC_is_visible produced wrong failure indication
Test failed
Abort
Kill Thread
```

So there aren't any configurations that stock gc-8.2.8 works on Haiku, it appears.

**Comment by waddlesplash** [comment]
Meanwhile, running my branch with `--disable-threads` (i.e. unmapping still enabled) works just fine:
```
Running on HAIKU/X86_64 target
Supported VDBs: manual default
Completed 6 tests
Allocated 3819281 collectable objects
Allocated 1218 uncollectable objects
Allocated 7754838 atomic objects
Reallocated 756 objects
Garbage collection after fork is tested too
Finalized 13218/13218 objects - finalization is probably OK
Total number of bytes allocated is 1111161792
Total memory use by allocated blocks is 184320 bytes
Final heap size is 184320 bytes
Obtained 9424896 bytes from OS (of which 8781824 bytes unmapped)
Final number of reachable objects is 3452
Full collections took 427 ms
World-stopped pauses took 352 ms (54 us each in avg.)
Completed 2047 collections
Collector appears to work (testing done in 1015 ms)
```

**Comment by waddlesplash** [comment]
Tried with `--disable-parallel-mark --disable-munmap` (i.e. threads still enabled) and it crashes constantly. So the problem really is just threads support it appears.

**Comment by waddlesplash** [comment]
Enabling a few debug prints, the collector does detect thread startups correctly:
```
Running on HAIKU/X86_64 target
Supported VDBs: manual default
Starting thread 0x11732d312af0, sp= 0x7f98ce9c77c0
Starting thread 0x11732d313cb0, sp= 0x7f9418937af0
Starting thread 0x11732d314e70, sp= 0x7f988d33d500
Starting thread 0x11732d316030, sp= 0x7f6ee470e240
GC Warning: Finalization cycle involving 0x1a1c99dea60
GC Warning: Finalization cycle involving 0x1a1c99dea60
...
```

**Comment by waddlesplash** [comment]
`GC_ASSERTIONS` caught something:
```
Running on HAIKU/X86_64 target
Supported VDBs: manual default
Starting thread 0x10c41ae305b0, sp= 0x7f920b15bea0
Starting thread 0x10c41ae31770, sp= 0x7fe5044ef2d0
Starting thread 0x10c41ae32930, sp= 0x7fe32c92d930
Starting thread 0x10c41ae33af0, sp= 0x7f6c43e2b270
Assertion failure: extra/../malloc.c:709
assertion failure
Assertion failure: extra/../malloc.c:709
assertion failure
```

**Comment by ivmai** [comment]
> `GC_ASSERTIONS` caught something:
> 
> ```
> Assertion failure: extra/../malloc.c:709
> ```

It means the lock is not held. What's the stack trace?

**Comment by waddlesplash** [comment]
Indeed, the problem is forking. With fork disabled, the test succeeds without any problems.

Poking around a bit, I think the issue is the current thread detection. If the main thread forks, then `pthread_self()` in the child will have the same value as before. But if some thread other than the main forks, `pthread_self()` will now report the main thread's address. We then remove all threads but the current (main) thread from the thread list, which is wrong because our stack address is actually of some child thread, and then things break.

I don't see any guarantees in the POSIX specification that the child thread `pthread_self()` will be the same as the parent application's `pthread_self()` before it was called? If you need a thread-specific identifier that survives across forks, a TLS slot would be the safest way to do this.

It may be possible to patch Haiku and make this work better, I suppose, but POSIX doesn't require us to do that, I think.

**Comment by waddlesplash** [comment]
Alternatively, you can just store the current thread ID in a global variable in `fork_prepare_proc` and make use of it in `fork_child_proc`, I suppose.

**Comment by waddlesplash** [comment]
https://github.com/ivmai/bdwgc/pull/678

**Comment by waddlesplash** [comment]
> Alternatively, you can just store the current thread ID in a global variable in fork_prepare_proc and make use of it in fork_child_proc, I suppose.

Yet another alternative: fetch the forked thread's stack base and use that to compare the threads.

**Comment by waddlesplash** [comment]
Well, any of them would work, but I don't know which you'd prefer. I'd probably pick the "check stack address" myself, but if (as on some platforms) finding the stack base address is already unreliable, perhaps that's not the best idea?

**Comment by waddlesplash** [comment]
I committed the patch so far to HaikuPorts in https://github.com/haikuports/haikuports/commit/9205cebac12f4574d6c34ca1e261929ecd60a269 so anyone on Haiku using boehm_gc isn't stuck waiting for the fixes anyway.
